### PR TITLE
 Reverse geocoding based on distance with coordinate

### PR DESCRIPTION
**Pull request by smadha**
#23

Two new parameters
1.  -r,--enable-reverse <true / false >           Add on indexing option for reverse geocoding. Defaults to false
2. -sr,--search-reverse <latitude , longitude>   Search locations near this coordinate

Sample commands-

```
lucene-geo-gazetteer -b /path/to/geonames/file.txt -r true -i ./geoIndex

lucene-geo-gazetteer -sr 34.022157 -118.285152
[{"name":"Los
Angeles","countryCode":"US","admin1Code":"CA","admin2Code":"037","latitu
de":34.05223,"longitude":-118.24368}]
```

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -185,19 +185,22 @@
...
 		<dependency>
 			<groupId>org.apache.lucene</groupId>
 			<artifactId>lucene-queryparser</artifactId>
 			<version>5.0.0</version>
 		</dependency>
-
+		<dependency>
+			<groupId>org.apache.lucene</groupId>
+			<artifactId>lucene-spatial</artifactId>
+			<version>5.0.0</version>
```
Comment (dsmiley) [inline pom.xml:201]: Why not latest in 5.x series?  or 6.x for that matter?

File: src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java [hunk 2]
```diff
@@ -102,8 +119,14 @@
 	private static Directory indexDir;
 	private static int hitsPerPage = 8;
 
-	private IndexReader indexReader;
+	//sort descending on population
+	SortField populationSort = new SortedNumericSortField(FIELD_NAME_POPULATION, SortField.Type.LONG, true);
```
Comment (dsmiley) [inline src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java:125]: Wouldn't distance be a better sort?

File: src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java [hunk 3]
```diff
@@ -155,6 +178,27 @@ public GeoNameResolver(String indexPath) throws IOException {
 
 	}
 
+	public List<Location> searchNearby(Double latitude, Double longitude, Double distance, String indexerPath, int count) throws IOException {
```
Comment (dsmiley) [inline src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java]: Naming a "distance" variable as-such without units leads to bugs in misunderstanding the units.  I suggest appending "Deg", or switch to "Km".

File: src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java [hunk 4]
```diff
@@ -155,6 +178,27 @@ public GeoNameResolver(String indexPath) throws IOException {
...
+		Sort sort = new Sort(populationSort);
+		TopDocs topDocs = searcher.search(new MatchAllDocsQuery(), filter, hitsPerPage, sort);
+
+		ScoreDoc[] scoreDocs = topDocs.scoreDocs;
+		HashMap<String, List<Location>> allCandidates = new HashMap<String, List<Location>>();
+
+		getMatchingCandidates(searcher, allCandidates, key, scoreDocs);
+		List<Location> results = allCandidates.get(key);
+		
+		return results.subList(0, count > results.size() ? results.size() : count);
```
Comment (dsmiley) [inline src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java]: If you only want as much as "count" docs back, then tell Lucene that in the searcher.search() method.

File: src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java (hunk 2 above)
Comment (smadha) [reply src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java:125]: Sort by population gives us more known locations first. We saw lot of less known locations in allCountries.txt like restaurants, parks, building etc. We started with a use case of finding cities so this sort will ensure we don't get noise. Also we have chosen a smaller distance (should make it configurable though) so we won't get multiple cities.

But you are right we should consider distance too. Later I want to do give some weight to distance as well like we do in ranking algorithm after string matching.

File: src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java (hunk 3 above)
Comment (smadha) [reply src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java]: Great suggestion. I will see how I can change unit to Km/mile

File: src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java (hunk 4 above)
Comment (smadha) [reply src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java]: You are right I did it wrong way. As we are not doing any ranking after we retrieve results from lucene there is no need of getting a sublist here.

Thanks for reviewing PR. Great suggestions !

File: pom.xml (hunk 1 above)
Comment (smadha) [reply pom.xml:201]: We are using 5.0 to be consistent with other lucene dependencies. If there is a need to upgrade we should upgrade all lucene dependencies

File: src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java (hunk 4 above)
Comment (smadha) [reply src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java]: Resolved

File: src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java (hunk 3 above)
Comment (smadha) [reply src/main/java/edu/usc/ir/geo/gazetteer/GeoNameResolver.java]: Resolved

**Pull request merged**